[OMEdit] Support animation of Vector visualizer

anotheruserofgithub opened the pull request:
### Introduction

This PR implements support for the animation of the Vector visualizer.
It involved quite a lot of refactoring to properly abstract all types of visualizers.
I'm sorry for that, but don't be afraid! Nothing has changed in terms of functionality.

### Description

[OMEdit] Refactoring in preparation for Vector

- Create abstract class AbstractVisualizerObject
- Make ShapeObject class inherit AbstractVisualizerObject
- Refactor a lot of things for proper abstraction of visualizer types
- Context menu of a visualizer:
  - Change icon of action "Reset Transparency and Texture"
    for it to be different from "Remove Texture"
  - Add a number picker for the user to "Change Transparency"
    instead of forcing the value to 50%
- Minor improvements and code cleaning

[OMEdit] Renaming in preparation for Vector

- Rename AbstractVisualizers -> AbstractVisualizer
- Rename Shapes -> Shape
- Rename Visualizer* -> Visualization*

[OMEdit] Implement visualization of Vector

- Create VectorObject class inheriting AbstractVisualizerObject
- Read VectorObject attributes from MAT/CSV/FMU files
- Visualize VectorObject as a composite of three OSG shapes
  (one cylinder for the shaft, two cones for both heads)
- Select scale/length/radius constants from MSL default parameters

### Code Review

- I tried to harmonize the way smart pointers `osg::ref_ptr<>` are used but I am not sure where exactly they are needed and where they can/should be removed.
- I wanted to test my modifications with CSV and FMU as well, but it seems [animation is not supported for these cases](https://github.com/OpenModelica/OpenModelica/blob/fad40625d8775401e6b24833b2a9b655726b6b40/OMEdit/OMEditLIB/Simulation/SimulationDialog.cpp#L1962), is that right? Is there some way to verify that I did not break anything?
- I made sure to be able to replicate the example `Modelica.Mechanics.MultiBody.Examples.Elementary.ForceAndTorque` by choosing the same scale factors as [the ones set by the model](https://github.com/modelica/ModelicaStandardLibrary/blob/1dd53117cde44b3b20fc35079b79f7ba34af0d5f/Modelica/Mechanics/MultiBody/Examples/Elementary/ForceAndTorque.mo#L4), but this is not a long-term behavior as we should rather select scales based on custom heuristics for a nice and informative animation (see modelica/modelica#3991 for a detailed discussion).
- I did not implement zoom-invariance yet (see modelica/modelica#3991 and [this comment](https://github.com/modelica/ModelicaStandardLibrary/issues/3931#issuecomment-1016347058) in modelica/modelica#3931), I guess this would not be difficult but I am not certain that we really want to do that since the latter issue is still open.

So to wrap up, vector scales are not determined very well for now, but I do not know how we would want to do that.
Anyway, users can always manually rescale vectors themselves by adapting the coordinates so that the animation meets their expectations, thus I do not think this is a blocking issue (obviously, different Modelica tools would produce different results but we cannot avoid that, as mentioned in modelica/modelica#3991).
Therefore I would suggest merging this PR if it is approved, and we can work on the scaling later.
If we wait too much it can become complicated to rebase the commits.

If @adeas31 or anybody could have a look and validate my changes, I would be grateful!
If this can help, I also pushed [my working branch](https://github.com/anotheruserofgithub/OpenModelica/tree/omedit_multibody_visualizers_vector_wip) with thirty-four smaller (dirty) commits.
I would have done more clean commits from the beginning if I knew it would require that much refactoring.
Do not hesitate to ping me if you have any doubt. ;)

### Related Issues

- TODO: Support missing Shape types (beam & gearwheel) as well as `extra` parameter, and draw a nicer spring
- MSL problems reported in modelica/modelica#3989 & modelica/modelica#3991
- Fixes #8211

adeas31 commented:
I will review this asap.

adeas31 commented:
@anotheruserofgithub can you please fix the cmake build.

adeas31 commented on OMEdit/OMEditLIB/Animation/ViewerWidget.cpp:
```diff
@@ -224,247 +224,260 @@ void ViewerWidget::mousePressEvent(QMouseEvent *event)
-  QAction action5(QIcon(":/Resources/icons/texture.svg"), tr("Apply Custom Texture"), this);
-  QAction action6(QIcon(":/Resources/icons/undo.svg"), tr("Remove Texture"), this);
+  AbstractVisualizerObject* visualizer = nullptr;
+  if ((visualizer = mpAnimationWidget->getVisualization()->getBaseData()->getVisualizerObjectByID(mSelectedVisualizer))) {
```
I can't the context menu to work. For me the `mSelectedVisualizer` is always empty because `computeIntersections` fails.

anotheruserofgithub replied:
My bad! This condition should not have been put there. This is fixed in c32e2fc.

adeas31 reviewed (changes requested):
Looks good.

I wonder why some of the files are not renamed and instead deleted and added as new e.g., `VisualizationCSV.cpp`. It should be same as `VisualizationCSV.h` otherwise the commit history is gone.

anotheruserofgithub commented:
FYI, I just added consecutive small commits 590ca8a, 33c54e5, c32e2fc, 2acc735 and amended bffb371 in consequence.

anotheruserofgithub commented:
> I wonder why some of the files are not renamed and instead deleted and added as new

Apparently this is due to squashing commits resulting in renamed files that have less than [50% similarity index](https://git-scm.com/docs/git-diff-index#Documentation/git-diff-index.txt--Mltngt).
I tried to use the git option `--find-renames` during rebase/merge/cherry-pick but as far as I understand, this only modifies how the diff is presented.

To avoid this deletion/addition behavior, for this PR I squashed 9d3e9fc..2acc735 into the first commit, f84b596..0b5fb1a into the second and f471fa0..d4884f2 into the third one.

So from the [Files changed](https://github.com/OpenModelica/OpenModelica/pull/9075/files) tab, if you show changes for a single commit, files are correctly indicated as renamed, but if you show all commits together, then there are deletions/additions.

anotheruserofgithub commented:
Is it possible to merge this PR without squashing its commits, or should we open and merge three separate PRs instead?
